### PR TITLE
New Module: Keycloak User Rolemapping

### PR DESCRIPTION
**Pull request by bratwurzt**
##### SUMMARY
Add `keycloak_user_rolemapping` module to provide management of direct assignment of client/realm roles to users. Also provides user role mapping for service-account users.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
`keycloak_user_rolemapping`

##### ADDITIONAL INFORMATION
Example:
```
- name: Map a client role to a user, authentication with credentials
  community.general.keycloak_user_rolemapping:
    realm: MyCustomRealm
    auth_client_id: admin-cli
    auth_keycloak_url: https://auth.example.com/auth
    auth_realm: master
    auth_username: USERNAME
    auth_password: PASSWORD
    state: present
    client_id: client1
    user_id: user1Id
    roles:
      - name: role_name1
        id: role_id1
      - name: role_name2
        id: role_id2
  delegate_to: localhost

- name: Map a client role to a service account user for a client, authentication with credentials
  community.general.keycloak_user_rolemapping:
    realm: MyCustomRealm
    auth_client_id: admin-cli
    auth_keycloak_url: https://auth.example.com/auth
    auth_realm: master
    auth_username: USERNAME
    auth_password: PASSWORD
    state: present
    client_id: client1
    service_account_user_client_id: clientIdOfServiceAccount
    roles:
      - name: role_name1
        id: role_id1
      - name: role_name2
        id: role_id2
  delegate_to: localhost
```

### CODE REVIEW

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_user_rolemapping
+
+short_description: Allows administration of Keycloak user_rolemapping with the Keycloak API
+
+version_added: 5.1.2
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:15]:
```suggestion
version_added: 5.3.0
```

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+                    - The unique identifier for this role_representation.
+                    - This parameter is not required for updating or deleting a role_representation but
+                      providing it will reduce the number of API calls required.
+
+extends_documentation_fragment:
+- community.general.keycloak
+
+
+author:
+    - Dušan Marković
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:109]:
```suggestion
    - Dušan Marković (@bratwurzt)
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some very first comments.

Please note that every new module must come with tests (integration or unit tests). Since other keycloak modules have unit tests as well, adding similar tests is probably a good idea.

**Comment by bratwurzt** [comment]
Could anyone help out on parameter naming? I'm quite bad at that.

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:4]: You need to add a copyright line, like this. Also the remainder of the license header needs to be adjusted.
```suggestion
# Copyright (c) 2022, ...
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later
```

File: tests/integration/targets/keycloak_user_rolemapping/aliases [hunk 4]
```diff
@@ -0,0 +1 @@
+unsupported
```
Comment (felixfontein) [inline tests/integration/targets/keycloak_user_rolemapping/aliases:1]:
```suggestion
# Copyright (c) 2022, ...
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

unsupported
```

File: tests/integration/targets/keycloak_user_rolemapping/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,140 @@
+---
+- name: Create realm
```
Comment (felixfontein) [inline tests/integration/targets/keycloak_user_rolemapping/tasks/main.yml:2]:
```suggestion
# Copyright (c) 2022, ...
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

- name: Create realm
```

File: tests/integration/targets/keycloak_user_rolemapping/vars/main.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+---
```
Comment (felixfontein) [inline tests/integration/targets/keycloak_user_rolemapping/vars/main.yml:1]:
```suggestion
---
# Copyright (c) 2022, ...
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

```

**Review by felixfontein — COMMENTED** [review]
~~Besides the following point, this looks good to me. I cannot really test this, but the code changes look good to me (i.e. they don't seem to break something that already worked) and since none of the maintainers reacted so far, I guess nobody objects. I'll merge this in a couple of days (assuming the below modification has been done).~~

Hmm, I must have done something strange, this should have ended up with a comment in **another** PR... I'll look at this one later, please ignore this for now :)

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2022, Dušan Marković (@bratwurzt)
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:4]:
```suggestion
# Copyright (c) 2022, Dušan Marković (@bratwurzt)
```

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_user_rolemapping
+
+short_description: Allows administration of Keycloak user_rolemapping with the Keycloak API
+
+version_added: 5.3.0
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:16]:
```suggestion
version_added: 5.7.0
```

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+    target_username:
+        type: str
+        description:
+            - Username of the user roles are mapped to.
+            - This parameter is not required (can be replaced by uid for less API call).
+
+    uid:
+        type: str
+        description:
+            - Id of the user to be mapped.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:63]:
```suggestion
            - ID of the user to be mapped.
```

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+            - This parameter is not required for updating or deleting the rolemapping but
+              providing it will reduce the number of API calls required.
+
+    service_account_user_client_id:
+        type: str
+        description:
+            - client_id of the service-account-user to be mapped.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:70]:
```suggestion
            - Client ID of the service-account-user to be mapped.
```
Or use `I(...)` to refer to an option name.

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+            - This parameter is not required for updating or deleting the rolemapping but
+              providing it will reduce the number of API calls required.
+
+    client_id:
+        type: str
+        description:
+            - Name of the client to be mapped (different than I(cid)).
+            - This parameter is required (can be replaced by cid for less API call).
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:78]:
```suggestion
            - This parameter is required if I(cid) is not provided (can be replaced by I(cid) to reduce the number of API calls that must be made).
```

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+
+    cid:
+        type: str
+        description:
+            - Id of the client to be mapped.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:83]:
```suggestion
            - ID of the client to be mapped.
```

File: plugins/modules/identity/keycloak/keycloak_user_rolemapping.py [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+    roles:
+        description:
+            - Roles to be mapped to the user.
+        type: list
+        elements: dict
+        suboptions:
+            name:
+                type: str
+                description:
+                    - Name of the role_representation.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_rolemapping.py:96]:
```suggestion
                    - Name of the role representation.
```
Same below.

File: tests/integration/targets/keycloak_user_rolemapping/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,143 @@
+# Copyright: (c) 2022, Dušan Marković (@bratwurzt)
```
Comment (felixfontein) [inline tests/integration/targets/keycloak_user_rolemapping/tasks/main.yml:1]:
```suggestion
# Copyright (c) 2022, Dušan Marković (@bratwurzt)
```

File: tests/integration/targets/keycloak_user_rolemapping/aliases [hunk 15]
```diff
@@ -0,0 +1,4 @@
+# Copyright: (c) 2022, Dušan Marković (@bratwurzt)
```
Comment (felixfontein) [inline tests/integration/targets/keycloak_user_rolemapping/aliases:1]:
```suggestion
# Copyright (c) 2022, Dušan Marković (@bratwurzt)
```

File: tests/integration/targets/keycloak_user_rolemapping/vars/main.yml [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+# Copyright: (c) 2022, Dušan Marković (@bratwurzt)
```
Comment (felixfontein) [inline tests/integration/targets/keycloak_user_rolemapping/vars/main.yml:2]:
```suggestion
# Copyright (c) 2022, Dušan Marković (@bratwurzt)
```

**Review by felixfontein — APPROVED** [review]
Since nobody else bothered to comment on this, and it looks good enough to me (I don't use keycloak so I cannot really comment on that aspect), I'm declaring this ready to merge and will merge it :)

**Comment by felixfontein** [comment]
@bratwurzt thanks a lot for your contribution, and sorry that it took so long to get it merged!

**Comment by bratwurzt** [comment]
> @bratwurzt thanks a lot for your contribution, and sorry that it took so long to get it merged!

No worries, glad I could contribute.

**Pull request merged**